### PR TITLE
[FLINK-12812][runtime] Set resource profiles for task slots

### PR DESCRIPTION
**Pull request by xintongsong**
## What is the purpose of the change

This pull request sets resource profiles for task slots according to resources the TaskExecutor is started with. This is one of the steps towards enabling fine grained resource management.

## Brief change log

- Add managed memory in ResourceSpec and ResourceProfile.
- Set resource profiles for task slots when starting TaskManagers.
- Set resource profiles for pending task manager slots.

## Verifying this change

This change added tests and can be verified as follows:

 - Added TaskExecutorTest#testSlotReportResourceProfile that validates that slot report created by the TaskExecutor contains correct task slot resource profiles.
- Added YarnResourceManagerTest#testCreateSlotsPerWorker that validates that slot resource profile calculated by TaskManagers (for slot tables) and ResourceManagers (for pending task manager slots) are the same.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java [hunk 1]
```diff
@@ -522,4 +534,68 @@ public void testOnContainerCompleted() throws Exception {
 			});
 		}};
 	}
+
+	/**
+	 * Test that RM and TM calculate same slot resource profile.
```
Comment (gaoyunhaii) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java]: Tests

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java [hunk 2]
```diff
@@ -163,4 +163,11 @@
 	boolean updatePartitionInfo(
 		ExecutionAttemptID consumerID,
 		PartitionInfo partitionInfo) throws IOException, InterruptedException;
+
+	/**
+	 * Get total memory size for network buffers in bytes.
```
Comment (gaoyunhaii) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: Gets

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 3]
```diff
@@ -1673,9 +1681,11 @@ public void testTerminationOnFatalError() throws Throwable {
 	@Test(timeout = 10000L)
 	public void testSyncSlotsWithJobMasterByHeartbeat() throws Exception {
 		final TaskSlotTable taskSlotTable = new TaskSlotTable(
-				Arrays.asList(ResourceProfile.UNKNOWN, ResourceProfile.UNKNOWN),
-				timerService);
-		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder().setTaskSlotTable(taskSlotTable).build();
+			Arrays.asList(ResourceProfile.UNKNOWN, ResourceProfile.UNKNOWN),
+			timerService);
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder().setTaskSlotTable(taskSlotTable)
+
```
Comment (gaoyunhaii) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: Remove this empty line.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/ResourceProfileTest.java [hunk 4]
```diff
@@ -31,10 +31,10 @@
...
 	@Test
 	public void testMatchRequirement() throws Exception {
-		ResourceProfile rp1 = new ResourceProfile(1.0, 100, 100, 100, 0, Collections.emptyMap());
-		ResourceProfile rp2 = new ResourceProfile(1.0, 200, 200, 200, 0, Collections.emptyMap());
-		ResourceProfile rp3 = new ResourceProfile(2.0, 100, 100, 100, 0, Collections.emptyMap());
-		ResourceProfile rp4 = new ResourceProfile(2.0, 200, 200, 200, 0, Collections.emptyMap());
+		ResourceProfile rp1 = new ResourceProfile(1.0, 100, 100, 100, 0, 0, Collections.emptyMap());
+		ResourceProfile rp2 = new ResourceProfile(1.0, 200, 200, 200, 0, 0, Collections.emptyMap());
+		ResourceProfile rp3 = new ResourceProfile(2.0, 100, 100, 100, 0, 0, Collections.emptyMap());
+		ResourceProfile rp4 = new ResourceProfile(2.0, 200, 200, 200, 0, 0, Collections.emptyMap());
```
Comment (gaoyunhaii) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/ResourceProfileTest.java]: Do we need to add a test for the newly added `approximate` method ？

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 5]
```diff
@@ -1744,6 +1757,39 @@ public void testSyncSlotsWithJobMasterByHeartbeat() throws Exception {
 		}
 	}
 
+	@Test
+	public void testSlotReportResourceProfile() throws Throwable {
+		Configuration config = new Configuration();
+		config.setInteger(TaskManagerOptions.NUM_TASK_SLOTS, 2);
+		config.setString(NettyShuffleEnvironmentOptions.NETWORK_BUFFERS_MEMORY_MIN, "100m");
```
Comment (gaoyunhaii) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: I think it may be better to define the test data in variables, and get the target value in the following `assertEquals` with the relationship between the variables, like
<pre>
    int networkMemoryInMB = 100;
    ....
    assertEquals(networkMemoryInMB / numSlots, resourceProfile.getNetworkMemoryInMB());
</pre>

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 6]
```diff
@@ -394,8 +394,10 @@ public void onContainersAllocated(List<Container> containers) {
 
 					try {
 						// Context information used to start a TaskExecutor Java process
+						// Use requested resource instead of allocated container resource, to make sure resource
```
Comment (gaoyunhaii) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: 1. Uses
2. A comma may be required for the last sentence.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 7]
```diff
@@ -585,7 +585,8 @@ private TaskManagerSlot createAndRegisterTaskManagerSlot(SlotID slotId, Resource
 	@Nullable
 	private PendingTaskManagerSlot findExactlyMatchingPendingTaskManagerSlot(ResourceProfile resourceProfile) {
 		for (PendingTaskManagerSlot pendingTaskManagerSlot : pendingSlots.values()) {
-			if (pendingTaskManagerSlot.getResourceProfile().equals(resourceProfile)) {
+			// there might be slight error between slot resource profiles calculated by RM / TM
+			if (pendingTaskManagerSlot.getResourceProfile().approximate(resourceProfile)) {
```
Comment (gaoyunhaii) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: I'm a little worry that the approximation may not be tolerant in TM side. For example, if the network memory of the available slot is just 1MB less than the request slot, it is very likely to cause the InsufficientBufferException in the TM side. 

I'm not sure why we can't make the request exactly the same with the available slot ? Or at least we must ensure pendingTaskManagerSlot.getResourceProfile().isMatching(resourceProfile) is also true.

File: flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java [hunk 8]
```diff
@@ -522,4 +534,68 @@ public void testOnContainerCompleted() throws Exception {
...
+		config2.setString(TaskManagerOptions.MANAGED_MEMORY_SIZE, "789m");
+		testCreateSlotsPerWorker(config2,  Resource.newInstance(800, 50));
+
+		Configuration config3 = new Configuration();
+		config3.setString(TaskManagerOptions.MANAGED_MEMORY_SIZE, "300m");
+		config3.setBoolean(TaskManagerOptions.MEMORY_OFF_HEAP, true);
+		testCreateSlotsPerWorker(config3,  Resource.newInstance(2000, 60));
+
+		Configuration config4 = new Configuration();
+		config4.setInteger(NettyShuffleEnvironmentOptions.NETWORK_NUM_BUFFERS, 50);
```
Comment (gaoyunhaii) [inline flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java]: I think it may be better to use the new NETWORK_BUFFERS_MEMORY_MIN/NETWORK_BUFFERS_MEMORY_MAX configuration instead.

**Comment by gaoyunhaii** [comment]
Thank @xintongsong for opening the PR! The PR looks good to me as a whole and I have left some comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 7 above)
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: The reason we need approximation is due to precision issue. Currently slot profile are calculated on both RM and TM, with the same rules but different implementations. As a result, the calculation results may have slight difference. We are planing to unify the memory calculation so that RM and TM use the same implementation in the next version. Once we have that, the approximation could be removed.

**Comment by xintongsong** [comment]
Thanks a lot for the review, @gaoyunhaii. I just addressed your comments.

**Comment by xintongsong** [comment]
@flinkbot attention @tillrohrmann

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java [hunk 9]
```diff
@@ -171,4 +171,11 @@ ShuffleIOOwnerContext createShuffleIOOwnerContext(
 	boolean updatePartitionInfo(
 		ExecutionAttemptID consumerID,
 		PartitionInfo partitionInfo) throws IOException, InterruptedException;
+
+	/**
+	 * Gets total memory size for network buffers in bytes.
+	 *
+	 * @return network memory size in bytes.
+	 */
+	long getMemorySize();
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: Actually we considered this issue in [FLINK-12110](https://issues.apache.org/jira/browse/FLINK-12110). 
I guess we might also need to know the memory type besides size, because we could not make an assumption of direct memory as now based on pluggable shuffle service.

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: My suggestion is not covering the network resources in `ResourceProfile` atm. As I know, the `ResourceProfile` from execution only covers managed memory from table api. If so we could also cover the managed memory to match as the first step, then focus on network resources future because it might involved in how to expose/calculate network resource on RM/JM side.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 10]
```diff
@@ -325,6 +321,40 @@ public static TaskManagerServices fromConfiguration(
...
+	private static TaskSlotTable createTaskSlotTable(
+			TaskManagerServicesConfiguration taskManagerServicesConfiguration,
+			long networkMemorySize,
+			long managedMemorySize) {
+
+		final int numOfSlots = taskManagerServicesConfiguration.getNumberOfSlots();
+		final long heapMemoryBytesPerSlot = (taskManagerServicesConfiguration.getMemoryType() == MemoryType.HEAP ?
+			taskManagerServicesConfiguration.getMaxJvmHeapMemory() - managedMemorySize :
+			taskManagerServicesConfiguration.getMaxJvmHeapMemory()) / numOfSlots;
+		final long networkMemoryBytesPerSlot = networkMemorySize / numOfSlots;
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: It might not cover network resource here as I suggested above.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java [hunk 11]
```diff
@@ -1191,8 +1197,32 @@ protected int getNumberRequiredTaskManagerSlots() {
 	//  Helper methods
 	// ------------------------------------------------------------------------
 
-	protected static Collection<ResourceProfile> createSlotsPerWorker(int numSlots) {
-		return Collections.nCopies(numSlots, ResourceProfile.ANY);
+	@VisibleForTesting
+	public static Collection<ResourceProfile> createSlotsPerWorker(Configuration config, long totalMemoryMB, int numSlots) {
+		final long cutoffMB = ContaineredTaskManagerParameters.calculateCutoffMB(config, totalMemoryMB);
+
+		final long networkMB = NettyShuffleEnvironmentConfiguration.calculateNetworkBufferMemory(
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: In reality `NettyShuffleEnvironmentConfiguration.calculateNetworkBufferMemory` should not be exposed outside based on pluggable shuffle service proposed in [FLINK-12110](https://issues.apache.org/jira/browse/FLINK-12110), so I suggest not covering network factor in `ResourceProfile` atm.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java (hunk 9 above)
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java]: Thanks for the input, @zhijiangW. I agree with you that we do not necessarily need matching network memory at the moment. I'll update the PR accordingly.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java (hunk 11 above)
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: I agree that we do not need to cover network memory in `ResourceProfile` in this version. However, I'm afraid we need to access `NettyShuffleEnvironmentConfiguration.calculateNetworkBufferMemory` on RM side anyway. Otherwise, we can not calculate managed memory size from the total container memory size. I think in the future we need a dedicated data structure to unify all the task executor resource calculations, exposing accesses of individual factors while hiding the detailed calculation steps.

**Comment by xintongsong** [comment]
Thank you for the review, @zhijiangW. I'll address your comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 12]
```diff
@@ -1621,6 +1621,11 @@ TaskExecutorToResourceManagerConnection getResourceManagerConnection() {
 		return resourceManagerHeartbeatManager;
 	}
 
+	@VisibleForTesting
+	SlotReport getSlotReport() {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: You can probably drop this method if you move the code that computes the slot resources into a separate method:
`computeSlotResourceProfile(taskManagerServicesConfiguration)` as suggested. Then you can test that logic in isolation and do not need access to the slot requests.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 13]
```diff
@@ -258,20 +259,13 @@ public static TaskManagerServices fromConfiguration(
 
 		// this call has to happen strictly after the network stack has been initialized
 		final MemoryManager memoryManager = createMemoryManager(taskManagerServicesConfiguration);
+		final long managedMemorySize = memoryManager.getMemorySize();
 
 		final BroadcastVariableManager broadcastVariableManager = new BroadcastVariableManager();
 
-		final List<ResourceProfile> resourceProfiles = new ArrayList<>(taskManagerServicesConfiguration.getNumberOfSlots());
```
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: This also allows for targeted testing of the resource computation code, which avoids many of the complexity of the new tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 14]
```diff
@@ -1729,14 +1738,16 @@ public void testSyncSlotsWithJobMasterByHeartbeat() throws Exception {
 			final AllocationID allocationIdOnlyInJM = new AllocationID();
 			final AllocationID allocationIdOnlyInTM = new AllocationID();
 
-			taskExecutorGateway.requestSlot(slotId1, jobId, allocationIdInBoth, "foobar", testingResourceManagerGateway.getFencingToken(), timeout);
-			taskExecutorGateway.requestSlot(slotId2, jobId, allocationIdOnlyInTM, "foobar", testingResourceManagerGateway.getFencingToken(), timeout);
+			taskExecutorGateway.requestSlot(slotId1, jobId, allocationIdInBoth, "foobar",
```
Comment (StephanEwen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: Lots of unrelated reformatting. Per Flink coding guides, these should be done in a change.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 15]
```diff
@@ -1845,6 +1860,42 @@ public void testSlotReportDoesNotContainStaleInformation() throws Exception {
 		}
 	}
 
+	@Test
+	public void testSlotReportResourceProfile() throws Throwable {
```
Comment (StephanEwen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: I think this test shows that the abstractions are too intermingled.

The code that was added in TaskExecutor was a change to replace the UNKNOWN profile attached to a slot by a profile with proper values. The code only takes the task manager's config (or managed memory size), and determines the size of the slots.

Such a change should not need the creation of a full TaskExecutor to test it. That means something is not right in the abstraction.

This test setup means that every time the TaskExecutor setup changes, the test for the simple code "config -> profile" needs to be adjusted, which contributes to making maintenance complex.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 7 above)
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: Why does the RM re-calculate the resource profile? This assumes that the RM and TM not only follow the same rules, but have the same config. That is actually not always the case, I have seen standalone setups where different TMs had different configs.

**Comment by xintongsong** [comment]
Thanks for the review, @StephanEwen. 

I would like to explain regarding your concern about the assumption of RM/TM having same configuration:

The reason we need to calculate TM's slot resource profiles on RM side is that, we need to set resource profile for `PendingTaskManagerSlot` before the corresponding TM is started. 

Currently, Flink can assign a pending slot to a slot request before the TM is started and registered. In this way, the subsequent slot requests will first consume slots on the pending TM (for multi-slot TMs) before requesting and launching a new one. When the TM is registered, the SlotManager matches the registered new slot to a `PendingTaskManagerSlot` with the same resource profile, and assigns the registered slot to the same slot request that the pending slot is assigned to (if any).

Before this PR, both the pending slot on RM side and the actual slot on TM side have the same resource profile `ANY`, which can be matched with the method `equals`. Since this PR sets the slot resource profile on TM side to the actual resource of the slot, we need to set the resource profile for the pending slots on RM side in the same way. This is way I introduced calculating TM's slot resource profiles on RM side, and the approximate matching.

The assigning over pending slots and the RM side slot resource calculating only happens on Yarn/Mesos. In these scenarios, TMs do have the same configuration as RM does, which is transmitted from RM side. For a standalone cluster, there should be no pending slots because RM can not actively start any TM.

Except for the `PendingTaskManagerSlot`, RM does use the slot resource profile reported from TM for matching slot request against registered slots, and converting requested `UNKNOWN` resource profile to a default value (as shown in the following PR #8846 for dynamic managed memory). Therefore, it should not cause problems on a standalone cluster with TMs having different configs.

It's my bad not making these clear in codes and comments. For the rest of your comments, I'll address them ASAP. I especially admire your suggestions on encapsulation and simplifying tests. It's a good lesson for me. Thank you again.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/ResourceSpec.java [hunk 16]
```diff
@@ -240,7 +254,8 @@ public String toString() {
 				", heapMemoryInMB=" + heapMemoryInMB +
 				", directMemoryInMB=" + directMemoryInMB +
 				", nativeMemoryInMB=" + nativeMemoryInMB +
-				", stateSizeInMB=" + stateSizeInMB + extend +
+				", stateSizeInMB=" + stateSizeInMB +
+				", managedMemoryInMB=" + managedMemoryInMB + extend +
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/api/common/operators/ResourceSpec.java]: I don't think there is any missing space. The string `extend` is either start with ", " or empty.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java [hunk 17]
```diff
@@ -284,12 +301,39 @@ else if (obj != null && obj.getClass() == ResourceProfile.class) {
...
 					Objects.equals(extendedResources, that.extendedResources);
 		}
 		return false;
 	}
 
+	public boolean approximate(ResourceProfile that) {
+		if (that == null) {
+			return false;
+		}
+		if (Math.abs(this.cpuCores - that.cpuCores) > 1e-6f) {
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: Yes, this is about the rounding errors during the profile calculation.

The calculation of profiles involves multiplying integer values with floating values. Rounding the floating value product to integer value will cause error. E.g., rounded result of `Total * Fraction` and `Total - Total * (1 - Fraction)` may be different, where `Total` is an integer value and `Fraction` is a floating value.

Each time we do such rounding, we may get an error with max value 1. Since there are two of such fraction based calculation (for managed memory and network memory), I set the max error allowed here to 2.

This approximate matching is only used for matching `PendingTaskManagerSlot` with slot registered from TM. It can be replaced with exact matching once we unify the TM resource configuration. After we unify the TM resource configurations, there will be no more resource calculations on TM side. For Yarn, RM will calculate the resource profile and pass the calculation result to the TM to be started, so the pending slots and the actual slots should have exact same profiles. For standalone, there shouldn't be any pending slots.

[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: I'll add comments for these explanations.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java [hunk 18]
```diff
@@ -301,7 +345,8 @@ public String toString() {
 			", heapMemoryInMB=" + heapMemoryInMB +
 			", directMemoryInMB=" + directMemoryInMB +
 			", nativeMemoryInMB=" + nativeMemoryInMB +
-			", networkMemoryInMB=" + networkMemoryInMB + resources +
+			", networkMemoryInMB=" + networkMemoryInMB +
+			", managedMemoryInMB=" + managedMemoryInMB + resources +
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: Same as above. The string `resources` is either start with ", " or empty.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: As we discussed offline, I removed this method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 7 above)
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: As we discussed offline, I reverted changes here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 19]
```diff
@@ -325,6 +319,38 @@ public static TaskManagerServices fromConfiguration(
 			.createShuffleEnvironment(shuffleEnvironmentContext);
 	}
 
+	private static TaskSlotTable createTaskSlotTable(
+			TaskManagerServicesConfiguration taskManagerServicesConfiguration,
+			long managedMemorySize) {
+
+		final int numOfSlots = taskManagerServicesConfiguration.getNumberOfSlots();
+		final long heapMemoryBytesPerSlot = (taskManagerServicesConfiguration.getMemoryType() == MemoryType.HEAP ?
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: As we discussed offline, I reverted this change.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 20]
```diff
@@ -358,14 +384,14 @@ private static MemoryManager createMemoryManager(
 			float memoryFraction = taskManagerServicesConfiguration.getMemoryFraction();
 
 			if (memType == MemoryType.HEAP) {
-				long freeHeapMemoryWithDefrag = taskManagerServicesConfiguration.getFreeHeapMemoryWithDefrag();
-				// network buffers allocated off-heap -> use memoryFraction of the available heap:
-				long relativeMemSize = (long) (freeHeapMemoryWithDefrag * memoryFraction);
+				long maxJvmHeapMemory = taskManagerServicesConfiguration.getMaxJvmHeapMemory();
```
[the remark this answers is not in the document]
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: As we discussed offline, I reverted this change.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/ResourceSpec.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-core/src/main/java/org/apache/flink/api/common/operators/ResourceSpec.java]: You are right, the delimiter is in the first loop so all is fine.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/ResourceProfile.java]: You are right!

**Comment by StephanEwen** [comment]
Thanks @xintongsong this looks really good now.

Will take this over and merge the PR.

There are some minor adjustments I would like to do (like de-duplicate configuration logic between Yarn/Mesos and and de-duplicate memory size logic between TaskManagerServices and ResourceManager) but I would apply these on the fly during the merge, due to the feature freeze deadline...

**Comment by StephanEwen** [comment]
I have applied the fixes here: https://github.com/StephanEwen/incubator-flink/tree/xintong

Trying to merge this, unfortunately battling with some tests that are very sensitive to rounding errors in the whole bytes/megabytes computation.

**Comment by StephanEwen** [comment]
Manually merged in c8b93b354d6d72ef3cacc9b0b9adcdb1df26e625
